modules/indexing/src/test/java/org/apache/ignite/internal/processors/database/WalDisabledDuringIndexRecreateTest.java

```diff
@@ -0,0 +1,308 @@
+    public boolean cacheGrps;
+
+    /** */
+    @Parameterized.Parameters(name = "cacheGroups={0}")
+    public static Iterable<Object> data() {
+        return Arrays.asList(true, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setGridLogger(testLog)
+            .setConsistentId(igniteInstanceName)
+            .setClusterStateOnStart(INACTIVE)
+            .setDataStorageConfiguration(new DataStorageConfiguration()
+                .setDefaultDataRegionConfiguration(new DataRegionConfiguration()
+                    .setPersistenceEnabled(true))
+                .setMaxWalArchiveSize(UNLIMITED_WAL_ARCHIVE));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    @Test
+    public void testDisabled() throws Exception {
+        WALPointer walStartPtr = createDataAndDeleteIndexBin();
+
+        awaitRebuild();
+
+        assertEquals(
+            (Long)3L,
+            countWalRecordsByTypes(wp -> wp.compareTo(walStartPtr) > 0).getOrDefault(BTREE_PAGE_INSERT, 0L)
+        );
+    }
+
+    /** */
+    @Test
+    public void testRestartInCaseOfFailure() throws Exception {
+        WALPointer walStartPtr = createDataAndDeleteIndexBin();
+
+        String treeName = BPlusTree.treeName(
+            new BinaryBasicIdMapper().typeId(TestVal.class.getName()) + "_"
+                + TestVal.class.getSimpleName().toUpperCase() + "_F0_IDX",
+            H2_TREE
+        );
+
+        try {
+            AtomicInteger errCntr = new AtomicInteger(10);
+
+            // Setting up wrapper that will fail on 10 put operation.
+            BPlusTree.testHndWrapper = (tree, hnd) -> {
+                if (!tree.name().equals(treeName))
+                    return hnd;
+
+                PageHandler<Object, BPlusTree.Result> delegate = (PageHandler<Object, BPlusTree.Result>)hnd;
+
+                return new PageHandler<BPlusTree.Get, BPlusTree.Result>() {
+                    @Override public BPlusTree.Result run(
+                        int cacheId,
+                        long pageId,
+                        long page,
+                        long pageAddr,
+                        PageIO io,
+                        Boolean walPlc,
+                        BPlusTree.Get arg,
+                        int lvl,
+                        IoStatisticsHolder statHolder
+                    ) throws IgniteCheckedException {
+                        if (arg instanceof BPlusTree.Put && errCntr.decrementAndGet() == 0)
+                            throw new IgniteCheckedException("Test error!");
```

Comment:
Proper explanation required.
Looks like you also must check errCntr.get()==0 at the final?